lib/ansible/modules/windows/win_file.ps1

```diff
@@ -20,51 +20,601 @@
+        [DllImport("kernel32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        static extern bool FindNextFileNameW(
+            IntPtr hFindStream,
+            ref uint stringLength,
+            StringBuilder linkName);
+
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/aa364413(v=vs.85).aspx
+        [DllImport("kernel32.dll", SetLastError = true)]
+        static extern bool FindClose(IntPtr hFindFile);
+
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/aa364996(v=vs.85).aspx
+        [DllImport("kernel32.dll")]
+        static extern bool GetVolumePathName(
+            string lpszFileName,
+            [Out] StringBuilder lpszVolumePathName,
+            uint cchBufferLength);
+
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/hh707081(v=vs.85).aspx
+        [DllImport("shlwapi.dll", CharSet = CharSet.Auto)]
+        static extern bool PathAppend(
+            [In, Out] StringBuilder pszPath,
+            string pszMore);
+
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/aa365488(v=vs.85).aspx
+        [DllImport("kernel32.dll", CharSet = CharSet.Unicode, SetLastError = true)]
         public static extern bool RemoveDirectory(string lpPathName);
 
-        public static void DeleteSymLink(string linkPathName) {
-            bool result = RemoveDirectory(linkPathName);
-            if (result == false)
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/aa363915(v=vs.85).aspx
+        [DllImport("kernel32.dll", CharSet = CharSet.Unicode, SetLastError = true)]
+        public static extern bool DeleteFile(string lpFileName);
+
+        public SymLink(string path)
+        {
+            FileAttributes attr = File.GetAttributes(path);
+            if ((attr & FileAttributes.ReparsePoint) == FileAttributes.ReparsePoint)
+            {
+                // We are dealing with a Symbolic Link, Junction Point or Mount Point
+                SetReparsePointInfo(path);
+            }
+            else
+            {
+                if ((attr & FileAttributes.Directory) == FileAttributes.Directory)
+                {
+                    // It is a Directory but not a link
+                    targets = new string[0];
+                    type = "";
+                }
+                else
+                {
+                    // File is potentially a Hard Link, check and set info
+                    SetHardLinkInfo(path);
+                }
+            }
+        }
+
+        public string[] Targets
+        {
+            get
+            {
+                return targets;
+            }
+        }
+
+        public string Type
+        {
+            get
+            {
+                return type;
+            }
+        }
+
+        // Will add the SeBackupPrivilege Token to the current Process (only call once)
+        public static void EnableSeBackupPrivilege()
+        {
+            bool success;
```

Comment:
No- the places where you've got custom messaging are good- just getting rid of the `success = SomeMethodCall(); if(!success)` pattern (collapsing that into `if(!SomeMethodCall()) throw`) since the utility of inspecting the `success` var in a debugger is pretty limited. Different story if it were an error code or something, but just as a bool, you know if you drop into the block, it's because it was false.